addons/mail/models/res_users.py

```diff
@@ -82,37 +82,36 @@ def create(self, vals_list):
         users = super(Users, self).create(vals_list)
         # Auto-subscribe to channels unless skip explicitly requested
         if not self.env.context.get('mail_channel_user_nosubscribe', False):
-            self.env['mail.channel'].search([('group_ids', 'in', users.groups_id.ids)])._subscribe_users()
+            self.env['mail.channel'].search([('group_ids', 'in', users.groups_id.ids)])._subscribe_users_automatically()
         return users
 
     def write(self, vals):
         write_res = super(Users, self).write(vals)
         if 'active' in vals and not vals['active']:
-            self._unsubscribe_from_channels()
+            self._unsubscribe_from_private_channels()
         sel_groups = [vals[k] for k in vals if is_selection_groups(k) and vals[k]]
         if vals.get('groups_id'):
             # form: {'group_ids': [(3, 10), (3, 3), (4, 10), (4, 3)]} or {'group_ids': [(6, 0, [ids]}
             user_group_ids = [command[1] for command in vals['groups_id'] if command[0] == 4]
             user_group_ids += [id for command in vals['groups_id'] if command[0] == 6 for id in command[2]]
-            self.env['mail.channel'].search([('group_ids', 'in', user_group_ids)])._subscribe_users()
+            self.env['mail.channel'].search([('group_ids', 'in', user_group_ids)])._subscribe_users_automatically()
         elif sel_groups:
-            self.env['mail.channel'].search([('group_ids', 'in', sel_groups)])._subscribe_users()
+            self.env['mail.channel'].search([('group_ids', 'in', sel_groups)])._subscribe_users_automatically()
         return write_res
 
     def unlink(self):
-        self._unsubscribe_from_channels()
+        self._unsubscribe_from_private_channels()
         return super().unlink()
 
-    def _unsubscribe_from_channels(self):
+    def _unsubscribe_from_private_channels(self):
         """ This method un-subscribes users from private mail channels. Main purpose of this
             method is to prevent sending internal communication to archived / deleted users.
             We do not un-subscribes users from public channels because in most common cases,
             public channels are mailing list (e-mail based) and so users should always receive
             updates from public channels until they manually un-subscribe themselves.
         """
-        self.mapped('partner_id.channel_ids').filtered(lambda c: c.public != 'public').write({
-            'channel_partner_ids': [Command.unlink(pid) for pid in self.mapped('partner_id').ids]
-        })
+        memberships = self.env['mail.channel.partner'].sudo().search([('partner_id', 'in', self.partner_id.ids)])
+        memberships.filtered(lambda m: m.channel_id.public != 'public').unlink()
```

Comment:
Purpose is to have technical methods managing members use the technical model using sudo and avoiding issues like active flag management, ... Using user.partner_id.channel_ids would lead to the same kind of search anyway on the m2m between partners and channels that is the mail_channel_partner table. Normally it leads to the same result when speaking of queries but is done as technical model level directly.

Ok to include the public part directly in search.